R/build-home-index.R

```diff
@@ -89,66 +112,35 @@ data_home_sidebar <- function(pkg = ".", call = caller_env()) {
     toc = data_home_toc(pkg)
   )
 
-  if (is.null(pkg$meta$home$sidebar$structure)) {
-    sidebar_html <- paste0(
-      purrr::compact(sidebar_components[default_sidebar_structure()]),
-      collapse = "\n"
-    )
-    return(sidebar_html)
-  }
-
-  custom <- pkg$meta$home$sidebar$components
-  sidebar_custom <- unwrap_purrr_error(purrr::map(
-    set_names(names2(custom)),
-    function(comp) {
-      data_home_component(
-        custom[[comp]],
-        error_pkg = pkg,
-        error_path = paste0("home.sidebar.components.", comp),
-        error_call = call
-      )
-    }
-  ))
-  sidebar_components <- utils::modifyList(sidebar_components, sidebar_custom)
-
-  config_check_list(
-    sidebar_components,
-    has_names = sidebar_structure,
-    error_pkg = pkg,
-    error_path = "home.sidebar.components",
-    error_call = call
-  )
-
-  sidebar_final_components <- purrr::compact(
-    sidebar_components[sidebar_structure]
-  )
+  needs_components <- setdiff(structure, names(default_components))
+  custom_yaml <- config_pluck_sidebar_components(pkg, needs_components, call = call)  
+  custom_components <- purrr::map(custom_yaml, function(x) {
+    sidebar_section(x$title, markdown_text_block(x$text))
+  })
+  components <- utils::modifyList(default_components, custom_components)
 
-  paste0(sidebar_final_components, collapse = "\n")
+  sidebar <- purrr::compact(components[structure])
+  paste0(sidebar, collapse = "\n")
 }
 
 # Update sidebar-configuration.Rmd if this changes
 default_sidebar_structure <- function() {
   c("links", "license", "community", "citation", "authors", "dev")
 }
 
-data_home_component <- function(component,
-                                error_pkg,
-                                error_path,
-                                error_call = caller_env()) {
-  title <- config_check_string(
-    component$title,
-    error_pkg = error_pkg,
-    error_path = paste0(error_path, ".title"),
-    error_call = error_call
-  )
-  text <- config_check_string(
-    component$text,
-    error_pkg = error_pkg,
-    error_path = paste0(error_path, ".text"),
-    error_call = error_call
-  )
-
-  sidebar_section(title, bullets = markdown_text_block(text))
+config_pluck_sidebar_components <- function(pkg, new_components, call = caller_env()) {
+  base_path <- "home.sidebar.components"
+  components <- config_pluck_list(pkg, base_path, has_names = new_components, call = call)
+  
+  for (name in names(components)) {
+    component <- components[[name]]
+    component_path <- paste0(base_path, ".", name)
+    
+    config_pluck_list(pkg, component_path, has_names = c("title", "text"), call = call)
```

Comment:
Yeah, that's right. It's a tiny bit sloppy, but it's easier than writing another function that only check, but takes the same specification as `config_pluck_`. 

That said, I'm not 100% happy with this API, so it might change again a bit in the future (but not in this PR).

Reply:
Ok, thanks!